compute_tools/src/lsn_lease.rs

```diff
@@ -0,0 +1,187 @@
+                config.password(storage_auth_token.clone());
+            } else {
+                info!("Storage auth token not set");
+            }
+            config
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Renews lsn lease periodically so static compute are not affected by GC.
+fn lsn_lease_bg_task(compute: Arc<ComputeNode>, lsn: Lsn) -> Result<()> {
+    loop {
+        let (tenant_id, timeline_id, configs) = {
+            let state = compute.state.lock().unwrap();
+
+            let spec = state.pspec.as_ref().expect("spec must be set");
+
+            let configs = postgres_configs_from_state(&state);
+            (spec.tenant_id, spec.timeline_id, configs)
+        };
+
+        let valid_until = acquire_lsn_lease_with_retry(tenant_id, timeline_id, lsn, &configs)
+            .with_context(|| {
+                format!("failed to get lsn lease for {tenant_id}/{timeline_id}@{lsn}")
+            })?;
+
+        let valid_duration = valid_until
+            .duration_since(SystemTime::now())
+            .unwrap_or(Duration::ZERO);
+
+        // Sleep for 60 seconds less than the valid duration but no more than half of the valid duration.
+        let sleep_duration = valid_duration
+            .saturating_sub(Duration::from_secs(60))
+            .max(valid_duration / 2);
+
+        info!(
+            "lsn_lease_request succeeded, sleeping for {} seconds",
+            sleep_duration.as_secs()
+        );
+        thread::sleep(sleep_duration);
+    }
+}
+
+/// Acquires lsn lease in a retry loop.
+fn acquire_lsn_lease_with_retry(
+    tenant_id: TenantId,
+    timeline_id: TimelineId,
+    lsn: Lsn,
+    configs: &[postgres::Config],
+) -> Result<SystemTime> {
+    const MAX_ATTEMPTS: usize = 10;
+    let mut attempts = 0;
+    let mut retry_period_ms = 500.0;
+
+    loop {
+        let result = try_acquire_lsn_lease(tenant_id, timeline_id, lsn, configs);
+        match result {
+            Ok(_) => {
+                return result;
+            }
+            Err(ref e) if attempts < MAX_ATTEMPTS => {
+                warn!("Failed to acquire lsn lease: {e} (attempt {attempts}/{MAX_ATTEMPTS})",);
+                thread::sleep(Duration::from_millis(retry_period_ms as u64));
+                retry_period_ms *= 1.5;
+            }
+            Err(_) => {
+                return result;
+            }
+        }
+        attempts += 1;
+    }
+}
+
+/// Tries to acquire an LSN lease through PS page_service API.
+fn try_acquire_lsn_lease(
+    tenant_id: TenantId,
+    timeline_id: TimelineId,
+    lsn: Lsn,
+    configs: &[postgres::Config],
+) -> Result<SystemTime> {
+    fn get_valid_until(
+        config: &postgres::Config,
+        tenant_shard_id: TenantShardId,
+        timeline_id: TimelineId,
+        lsn: Lsn,
+    ) -> Result<SystemTime> {
+        let mut client = config.connect(NoTls)?;
+        let cmd = format!("lease lsn {} {} {} ", tenant_shard_id, timeline_id, lsn);
```

Comment:
Is there anything in the pageserver `lease lsn` API that can signal the caller that it's impossible to obtain the lease?

This is actually a broader question about the retry mechanism. If I calculate it right, 10 attempts with current backoff are slightly under 1 min, and we have seen in the past some pageservers being unresponsive >1 min due to restart, iirc

I'd be much more confident in a simpler yet persistent logic -- just catch and retry any errors indefinitely, BUT only if it's not a permanent error, i.e. we already raced with GC, so acquiring the lease is impossible.

Reply:
> Is there anything in the pageserver lease lsn API that can signal the caller that it's impossible to obtain the lease?

I added a mechanism on the pageserver side that returns `NULL` if the caller is impossible to obtain the lease (LSN already below gc_cutoff at the first request).

> I'd be much more confident in a simpler yet persistent logic -- just catch and retry any errors indefinitely, BUT only if it's not a permanent error, i.e. we already raced with GC, so acquiring the lease is impossible.

Yep, I agree. Switched to this approach in 1b26db464dd89130fac2c903ff78f003ca28da1c. I do wonder what's the best way to deal with permanent failure.  We wait for postgres forever unless we received a `/terminate` HTTP request.

Reply:
> I do wonder what's the best way to deal with permanent failure. We wait for postgres forever unless we received a /terminate HTTP request.

I think it's probably fine. I suggested some changes to error messages, so they were more explicit. We can watch how many errors we will really have

Other options:
1. Crash the whole compute, then it will be restarted and get stuck in the retry loop. Eventually, it should be suspended by cplane
2. Have some feedback channel to cplane to signal that we are unhealthy and must be terminated

Both require more work, so I wouldn't do it in the same PR. Instead just watch errors closer, once we enable static endpoints back

Reply:
> 2. Have some feedback channel to cplane to signal that we are unhealthy and must be terminated

Does cplane now consume the error in the `/status` response? We could use that.

Yes, this should be implemented in a separate PR.

Reply:
It does, but it currently expects to see it there only during the compute startup process. There is an activity monitor in cplane that pings all computes regularly, we can teach it to look at the error as well